### PR TITLE
`ultralytics 8.1.39` add YOLO-World training

### LINKED ISSUE #8156: Add YOLO World training
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

Related to #8039 and #8054

- [ ] add code for training for YOLO World models
- [ ] add validation code as well

### Use case

Language fine-tuning with corresponding image data

### Additional

- Establish requirements and/or features to be included

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by Laughing-q**
Usage:
```python
from ultralytics import YOLOWorld

model = YOLOWorld("yolov8s-worldv2.pt")
model.train(data="coco128.yaml")
```

TODO:
- [x] LVIS docs
- [x] Verify `WorldTrainerFromScratch`

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Enhancements in LVIS dataset support, model training updates, and technical refinements across the Ultralytics framework.

### 📊 Key Changes
- Added support and detailed documentation for the LVIS dataset.
- Updated CLIP model installation to use the Ultralytics repository.
- Introduced refactorings in model training for the YOLO-World architecture, including enhanced CLIP model support.
- Refined data augmentation processes, particularly for multi-modal (image + text) training scenarios.
- Made improvements in dataset management, including better cache file handling and support for additional datasets.
- Adjustments in model and dataset configurations for more accurate training results.

### 🎯 Purpose & Impact
- **Enhanced Dataset Support**: Including LVIS enhances the model's ability to train on a wider range of object categories, improving versatility and accuracy.
- **Model Training and Evaluation Improvements**: Updates in model training, especially with the introduction of YOLO-World-related features, pave the way for more advanced multi-modal learning capabilities. 
- **Robust Data Handling**: Improved dataset caching and configurations streamline the data preparation process, making model training more efficient.
- **Technical Refinements**: General code enhancements contribute to a more robust, efficient, and flexible machine learning pipeline.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@Laughing-q nice work with https://github.com/ultralytics/CLIP and fixing the pkg_resources warnings!!

**Comment by Laughing-q** [comment]
train_batch1

**Comment by Laughing-q** [comment]
@glenn-jocher PR is ready to review!

**Comment by Laughing-q** [comment]
@glenn-jocher ok now I'm pretty sure that training on grounding datasets(`GQA` and `flickr30k`) works properly and the results are correct to me. 
I trained a `yolov8s-worldv2` model from scratch on `GQA` and `flickr30k` for 25 epochs, then validated it on LVIS minival set and got mAP 14.4:
image
Noticed the official `yolov8s-worldv2` model got mAP 22.2 which was trained on datasets `object365v1`, `GQA` and `flickr30k` for 100 epochs.

**Comment by glenn-jocher** [comment]
@Laughing-q awesome, nice work here!

**Comment by glenn-jocher** [comment]
@Laughing-q hey I looked through this, this is really nice and comprehensive, including added docs and reference section pages.

I added a few missing docstrings and updated the val.py class_map, but that's the only changes I've made. Are there are sections you're not sure about, or is this ready to merge to main now?

**Comment by Laughing-q** [comment]
@glenn-jocher Thanks for reviewing! I updated the class_map to start from index 1 because that's what LVIS data needs when I intended to save json and manually validated mAP by using their API. Then I figured that for user cases they might not really care about whether the index starts from 0 or 1 so I eventually pushed the update. Do you think we should keep the index starting from 0? then probably create a class_map for LVIS just like we did for COCO?

**Comment by Laughing-q** [comment]
@glenn-jocher rest of the PR I think is all good, but since you were asking, I feel like it's better to hold it for another several hours and let me check if there're any places to improve, since it's a PR adding more than 2000 lines(1200 lines are from lvis.yaml though..). What do you think?

**Comment by Laughing-q** [comment]
@glenn-jocher ok I've added lvis api support to evaluate the final results just like we did for coco. Also I add an extra `+1` for category_id for lvis dataset while saving predictions in json format so it can be evaluated correctly. I think the PR is all set! :)

**Comment by glenn-jocher** [comment]
@Laughing-q PR merged!!

**Comment by ccl-private** [comment]
@Laughing-q Hi, can you help me take a look at this issue? https://github.com/ultralytics/ultralytics/issues/13793
